src/parser_util.rs

```diff
@@ -14,34 +15,222 @@ where
+            field_b.name.as_ref(),
+            &type_name
+        ));
+    };
+
+    has_same_type_shape(
+        &alias_or_name(field_a),
+        type_name,
+        &_field_a.type_,
+        &_field_b.type_,
+    )?;
+
+    if field_a.name != field_b.name {
+        return Err(format!(
+            "Fields '{}' conflict because '{}' and '{}' are different fields",
+            alias_or_name(field_a),
+            field_a.name.as_ref(),
+            field_b.name.as_ref(),
+        ));
+    }
+
+    for (arg_a_name, arg_a_value) in field_a.arguments.iter() {
+        let arg_b_value = field_b.arguments.iter().find_map(|(name, value)| {
+            if name == arg_a_name {
+                Some(value)
+            } else {
+                None
+            }
+        });
+        let args_match = match arg_b_value {
+            None => false,
+            Some(arg_b_value) => arg_b_value == arg_a_value,
+        };
+        if !args_match {
+            return Err(format!(
+                "Fields '{}' conflict because they have differing arguments",
+                alias_or_name(field_a),
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+pub fn has_same_type_shape(
+    field_name: &str,
+    type_name: &str,
+    type_a: &__Type,
+    type_b: &__Type,
+) -> Result<(), String> {
+    let mut type_a = type_a;
+    let mut type_b = type_b;
+
+    if matches!(type_a, __Type::NonNull(_)) || matches!(type_b, __Type::NonNull(_)) {
+        if let (__Type::NonNull(nullable_type_a), __Type::NonNull(nullable_type_b)) =
+            (type_a, type_b)
+        {
+            type_a = nullable_type_a.type_.as_ref();
+            type_b = nullable_type_b.type_.as_ref();
+        } else {
+            return Err(format!(
+                "Fields '{}' conflict because only one is non nullable",
+                field_name,
+            ));
+        }
+    }
+
+    if matches!(type_a, __Type::List(_)) || matches!(type_b, __Type::List(_)) {
+        if let (__Type::List(list_type_a), __Type::List(list_type_b)) = (type_a, type_b) {
+            type_a = list_type_a.type_.as_ref();
+            type_b = list_type_b.type_.as_ref();
+        } else {
+            return Err(format!(
+                "Fields '{}' conflict because only one is a list type",
+                field_name,
+            ));
+        }
+
+        return has_same_type_shape(field_name, type_name, type_a, type_b);
+    }
+
+    if matches!(type_a, __Type::Enum(_))
+        || matches!(type_b, __Type::Enum(_))
+        || matches!(type_a, __Type::Scalar(_))
+        || matches!(type_b, __Type::Scalar(_))
+    {
+        return if type_a == type_b {
+            Ok(())
+        } else {
+            Err(format!(
+                "Fields '{}' conflict due to mismatched types",
+                field_name,
+            ))
+        };
+    }
+
+    // TODO handle composite types?
+
+    // Subfield type shapes will be checked on a later pass.
+    Ok(())
+}
```

Comment:
I think this is confirming that the `__Type` of an and b are the same. As I understand, they need to be the exact same `__Type` in order to merge, correct?

If so, could we do something like make `__Type` hashable or equatable so they can be directly compared without this logic? The reason I prefer that kind of solution is because it would fail loudly if we introduce new `__Types` vs quietly not working with merging

Reply:
Yeah, this is a check to make sure both types have the same shape following the GQL specification linked in the discussion thread.

I did this explicitly because it gives me the ability to return specific errors for the different type conflicts, and keep the system close to spec.

I'm not sure if we should consider all `__Type` variants in the sameness check since it's not in spec, but an `Eq` trait would imply that the values are equal across across all variants. 

However, I do agree it should be co-located closer to the `__Type` definition.

Another thing I just noticed was that I forgot to include `type_name` in the error messages so it's an unused variable.

Reply:
> I'm not sure if we should consider all __Type variants in the sameness check since it's not in spec, but an Eq trait would imply that the values are equal across across all variants.

That's interesting, I didn't quite follow. Could I hassle you to give an example that highlights the issue?

Reply:
My guess was if a `__Type::Edge` and `__Type::Node` are compared, they should not fail the shape sameness check yet, as the `__Type::Node` might have a compatible set of fields for a selection to be merged with the `__Type::Edge`.

I think this also mainly stems from the specificity of the GraphQL merging spec in that I wanted to avoid adding false negatives to the sameness checks.